### PR TITLE
Fix unevaluated printing of negative sum in numerator

### LINKED ISSUE #24108: Incorrect printing of parse_latex('\\frac {-1 - 1}{2}')
I've just uncovered a bug when displaying LaTeX expressions involving numerators of the form `-X - Y` or `-X + func(Y)` in combination with purely numerical denominators.

Here's a relevant code snippet that demonstrates the problem: 
```
from sympy.parsing.latex import parse_latex
from sympy import srepr

expr = parse_latex('\\frac {-1 - 1}{2}')

print(expr)
print(srepr(expr))
```

Under `sympy 1.11.1`, the code above produces the following output:
```
-1 - 1*1/2
Mul(Pow(Integer(2), Integer(-1)), Add(Integer(-1), Mul(Integer(-1), Integer(1))))
```

This is clearly a printing problem as the string representation is missing proper parentheses around the `-1 - 1*1` numerator. (The `srepr` is fine, though.)

From what I can tell, the printing seems to fail when we pass `evaluate=False` to `sympy.Mul` in combination with certain arguments. For example, the above is equivalent to:
```
sympy.Mul(sympy.S('-1 - 1', evaluate=False), sympy.Pow(2, -1, evaluate=False), evaluate=False)
```
which again displays the same incorrect string, i.e. `-1 - 1*1/2`.

However, if we change the argument of `sympy.Pow` (i.e. we change the original denominator of the fraction) to an expression like `x`, then the printing is absolutely fine. Indeed, 
```
sympy.Mul(sympy.S('-1 - 1', evaluate=False), sympy.Pow(sympy.S('x'), -1, evaluate=False), evaluate=False)
``` 
displays `(-1 - 1*1)/x`, as desired.

Note that the same bug happens when we replace the numerator with expressions like `-1 + \\sqrt{3}`.

Lastly, here is a list of correct/incorrect string representations that might be helpful during debugging:
```
print(parse_latex('\\frac {-1 + 1}{2}'))
(-1 + 1)/2 ✅

print(parse_latex('\\frac {-1 - 1}{2}'))
-1 - 1*1/2 ❌
```

```
print(parse_latex('\\frac {-1 - 1}{200 + 0}'))
(-1 - 1*1)/(0 + 200) ✅

print(parse_latex('\\frac {-1 - 1}{200}'))
-1 - 1*1/200 ❌
```

```
print(parse_latex('\\frac {0 -x - y}{2}'))
(-y - (x + 0))/2 ✅

print(parse_latex('\\frac {-x - y}{2}'))
-x - y/2 ❌
```

```
print(parse_latex('\\frac {x + \\sqrt{y}}{2}'))
(x + sqrt(y))/2 ✅

print(parse_latex('\\frac {-x + \\sqrt{y}}{2}'))
-x + sqrt(y)/2 ❌
```

### PR DESCRIPTION
**Pull request by oscargus**
#### References to other Issues or PRs

Fixes #24108 

#### Release Notes

* printing
   * Fix str of numerator consisting of a sum where a minus signs can be extracted.

### CODE REVIEW

**Comment by oscargus** [comment]
Test program 
```
from sympy import * 
from sympy.parsing.latex import parse_latex
expr = parse_latex('\\frac {-1 - 1}{2}')
expr2 = parse_latex('\\frac {-\\sin(x)}{2}')
expr3 = parse_latex('\\frac {-a - b}{2}')
expr4 = parse_latex('\\frac {-1 - 1*3}{2}')
expr5 = parse_latex('\\frac {-1 - 1*1}{2}')
print(expr)
print(expr2)
print(expr3)
print(expr4)
print(expr5)
```

With this PR:
```
(-1 - 1*1)/2
-sin(x)/2
(-a - b)/2
(-3 - 1)/2
(-1 - 1)/2
```

With master:
```
-1 - 1*1/2
-sin(x)/2
-a - b/2
-3 - 1/2
-1 - 1/2
```

File: sympy/printing/tests/test_str.py [hunk 1]
```diff
@@ -286,6 +286,11 @@ def test_Mul():
     # issue 21537
     assert str(Mul(x, Pow(1/y, -1, evaluate=False), evaluate=False)) == 'x/(1/y)'
 
+    # Issue 24108
+    from sympy.core.parameters import evaluate
+    with evaluate(False):
+        str(Mul(Pow(Integer(2), Integer(-1)), Add(Integer(-1), Mul(Integer(-1), Integer(1))))) == "(-1 - 1*1)/2"
+        str(Mul(Pow(Integer(2), Integer(-1)), Mul(Integer(-1), sin(Symbol('x'))))) == '-sin(x)/2'
```
Comment (oscarbenjamin) [inline sympy/printing/tests/test_str.py:293]:
```suggestion
        assert str(Mul(Pow(Integer(2), Integer(-1)), Add(Integer(-1), Mul(Integer(-1), Integer(1))))) == "(-1 - 1*1)/2"
        assert str(Mul(Pow(Integer(2), Integer(-1)), Mul(Integer(-1), sin(Symbol('x'))))) == '-sin(x)/2'
```

**Comment by oscarbenjamin** [comment]
Looks good.

**Comment by smichr** [comment]
see also #24022

**Comment by oscarbenjamin** [comment]
Looks like the added tests fail.

**Comment by oscargus** [comment]
Fixed the tests and replaced `str` with `self._print` as @asmeurer correctly pointed out in https://github.com/sympy/sympy/issues/24108#issuecomment-1268960434

As @smichr points out, there are related issues. The test here must be updated when the -1*1 print issue is fixed.

**Comment by oscargus** [comment]
Not sure. The "ironic" thing is that `parse_latex('\\frac {-1 - 1*1}{2}')` is printed as `(-1 - 1)/2` but `parse_latex('\\frac {-1 - 1}{2}')` is printed as `(-1 - 1*1)/2`.

> The question coming back from there is why do you have a -1*1 expression in the first place?

I'm inclined to ask why anyone would like unevaluated printing at all...

**Comment by oscargus** [comment]
I think it was a parsing bug and should be fixed by #24096

**Comment by oscarbenjamin** [comment]
> I'm inclined to ask why anyone would like unevaluated printing at all...

I think what is really wanted is to be able to control how expressions are printed. In other words for some uses it should be possible to define an expression that is to be printed in a certain way and then have it printed in exactly that way.

This is sometimes for prettiness purposes e.g. if you aren't really doing much heavy computing with SymPy but you want it to produce displayed equations. Then this kind of thing ruins your equations:
```
In [6]: pdf = 1/(sigma*sqrt(2*pi))*exp(-((x - mu)/sigma)**2)

In [7]: pdf
Out[7]: 
             2 
    -(-μ + x)  
    ───────────
          2    
         σ     
√2⋅ℯ           
───────────────
     2⋅√π⋅σ
```
In that context someone might use `evaluate=False` to stop evaluation from messing up their expressions but then they still have to contend with the printer not printing their expressions in the intended way.

The other context is things related to code generation e.g. in floating point you might need to care about the distinction between `(x + y) + z` and `x + (y + z)`.

The problem is that SymPy just hasn't been designed with these uses in mind and `evaluate=False` is essentially a bolt on that never completely works.

**Comment by oscarbenjamin** [comment]
> I think it was a parsing bug and should be fixed by #24096

Yes, looks like it:
```python
In [2]: from sympy.parsing.latex import parse_latex

In [3]: parse_latex('\\frac {-1 - 1*1}{2}')
Out[3]: 
-1 - 1⋅1
────────
   2    

In [4]: parse_latex('\\frac {-1 - 1}{2}')
Out[4]: 
-1 - 1
──────
  2  
```

**Comment by asmeurer** [comment]
> In that context someone might use evaluate=False to stop evaluation from messing up their expressions but then they still have to contend with the printer not printing their expressions in the intended way.

That seems fine, albeit annoying. But many of the annoyances are due to things we shouldn't be doing in the printers anyway (like sorting). Again, if the printers could actually tell if an expression was "unevaluated" it would make this easier.

> The other context is things related to code generation e.g. in floating point you might need to care about the distinction between (x + y) + z and x + (y + z).

I don't like this as a use-case. If you want this sort of thing for code generation you should use special codegen nodes. evaluate=False is much more fragile for code generation because you're more likely to actually do computation with expressions that are used for codegen, but you never want to do computation on unevaluated expressions. That's where things break and you can get all sorts of bugs, because no other part of the codebase is designed around them. I know the original issue author mentioned code generation, but I don't see why you'd need `-1 - 1` to be unevaluated for codegen purposes.

**Pull request merged**